src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php

```diff
@@ -112,35 +112,51 @@ public function handle(GetResponseEvent $event)
      */
     private function attemptSwitchUser(Request $request)
     {
-        $token = $this->securityContext->getToken();
-        $originalToken = $this->getOriginalToken($token);
-
-        if (false !== $originalToken) {
-            if ($token->getUsername() === $request->get($this->usernameParameter)) {
-                return $token;
-            } else {
-                throw new \LogicException(sprintf('You are already switched to "%s" user.', $token->getUsername()));
-            }
+        if (null !== $this->logger) {
+            $this->logger->info(sprintf('Attempt to switch to user "%s"', $request->get($this->usernameParameter)));
         }
 
-        if (false === $this->accessDecisionManager->decide($token, array($this->role))) {
+        // token of the currently authenticated user
+        $sourceToken = $this->securityContext->getToken();
```

Comment:
don't know if `$sourceToken` is the best name, I choose it because the second parameter of new `SwitchUserRole('ROLE_PREVIOUS_ADMIN', $this->securityContext->getToken());` is called `$source`